File: backend/TirthSeva.API/Controllers/TemplesController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using TirthSeva.API.Models;
using TirthSeva.API.Services;

namespace TirthSeva.API.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class TemplesController : ControllerBase
    {
        private readonly TempleService _templeService;

        public TemplesController(TempleService templeService)
        {
            _templeService = templeService;
        }

        // Public endpoints - anyone can view temples
        [HttpGet]
        public async Task<ActionResult<List<Temple>>> GetAll()
        {
            var temples = await _templeService.GetAllTemplesAsync();
            return Ok(temples);
        }

        [HttpGet("{id}")]
        public async Task<ActionResult<Temple>> GetById(int id)
        {
            var temple = await _templeService.GetTempleByIdAsync(id);
            
            if (temple == null)
            {
                return NotFound();
            }

            return Ok(temple);
        }

        [HttpGet("search")]
        public async Task<ActionResult<List<Temple>>> Search([FromQuery] string? query, [FromQuery] string? state, [FromQuery] string? city)
        {
            var temples = await _templeService.SearchTemplesAsync(query, state, city);
            return Ok(temples);
        }

        [HttpGet("states")]
        public async Task<ActionResult<List<string>>> GetStates()
        {
            var states = await _templeService.GetStatesAsync();
            return Ok(states);
        }

        [HttpGet("cities/{state}")]
        public async Task<ActionResult<List<string>>> GetCities(string state)
        {
            var cities = await _templeService.GetCitiesByStateAsync(state);
            return Ok(cities);
        }

        // Provider endpoints - requires ServiceProvider role
        [Authorize(Roles = "ServiceProvider")]
        [HttpGet("my-temples")]
        public async Task<ActionResult<List<Temple>>> GetMyTemples()
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var temples = await _templeService.GetProviderTemplesAsync(userId);
            return Ok(temples);
        }

        [Authorize(Roles = "ServiceProvider")]
        [HttpPost]
        public async Task<ActionResult<Temple>> Create([FromBody] CreateTempleRequest request)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            
            var temple = new Temple
            {
                ServiceProviderId = userId,
                Name = request.Name,
                Location = request.Location,
                City = request.City,
                State = request.State,
                Description = request.Description,
                ImagePath = request.ImagePath,
                Latitude = request.Latitude,
                Longitude = request.Longitude
            };

            var created = await _templeService.CreateTempleForProviderAsync(temple);
            return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
        }

        [Authorize(Roles = "ServiceProvider")]
        [HttpPut("{id}")]
        public async Task<ActionResult<Temple>> Update(int id, [FromBody] UpdateTempleRequest request)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            
            var temple = new Temple
            {
                Id = id,
                ServiceProviderId = userId,
                Name = request.Name,
                Location = request.Location,
                City = request.City,
                State = request.State,
                Description = request.Description,
                ImagePath = request.ImagePath,
                Latitude = request.Latitude,
                Longitude = request.Longitude
            };

            var updated = await _templeService.UpdateTempleForProviderAsync(id, temple, userId);
            
            if (updated == null)
            {
                return NotFound(new { message = "Temple not found or you don't have permission to update it" });
            }

            return Ok(updated);
        }

        [Authorize(Roles = "ServiceProvider")]
        [HttpDelete("{id}")]
        public async Task<ActionResult> Delete(int id)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var result = await _templeService.DeleteTempleForProviderAsync(id, userId);
            
            if (!result)
            {
                return NotFound(new { message = "Temple not found or you don't have permission to delete it" });
            }

            return NoContent();
        }
    }

    public class CreateTempleRequest
    {
        public string Name { get; set; } = string.Empty;
        public string Location { get; set; } = string.Empty;
        public string City { get; set; } = string.Empty;
        public string State { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public string? ImagePath { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }
    }

    public class UpdateTempleRequest
    {
        public string Name { get; set; } = string.Empty;
        public string Location { get; set; } = string.Empty;
        public string City { get; set; } = string.Empty;
        public string State { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public string? ImagePath { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }
    }
}

File: backend-spring/src/main/java/com/tirthseva/api/controller/TemplesController.java
package com.tirthseva.api.controller;

import com.tirthseva.api.dto.temple.CreateTempleRequest;
import com.tirthseva.api.dto.temple.UpdateTempleRequest;
import com.tirthseva.api.entity.Temple;
import com.tirthseva.api.entity.User;
import com.tirthseva.api.service.TempleService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/temples")
@RequiredArgsConstructor
public class TemplesController {

    private final TempleService templeService;

    @GetMapping
    public ResponseEntity<List<Temple>> getAll() {
        List<Temple> temples = templeService.getAllTemples();
        return ResponseEntity.ok(temples);
    }

    @GetMapping("/{id}")
    public ResponseEntity<Temple> getById(@PathVariable Integer id) {
        Temple temple = templeService.getTempleById(id);

        if (temple == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(temple);
    }

    @GetMapping("/search")
    public ResponseEntity<List<Temple>> search(
            @RequestParam(required = false) String query,
            @RequestParam(required = false) String state,
            @RequestParam(required = false) String city) {
        List<Temple> temples = templeService.searchTemples(query, state, city);
        return ResponseEntity.ok(temples);
    }

    @GetMapping("/states")
    public ResponseEntity<List<String>> getStates() {
        List<String> states = templeService.getStates();
        return ResponseEntity.ok(states);
    }

    @GetMapping("/cities/{state}")
    public ResponseEntity<List<String>> getCities(@PathVariable String state) {
        List<String> cities = templeService.getCitiesByState(state);
        return ResponseEntity.ok(cities);
    }

    @PreAuthorize("hasRole('ServiceProvider')")
    @GetMapping("/my-temples")
    public ResponseEntity<List<Temple>> getMyTemples(@AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        List<Temple> temples = templeService.getProviderTemples(user.getId());
        return ResponseEntity.ok(temples);
    }

    @PreAuthorize("hasRole('ServiceProvider')")
    @PostMapping
    public ResponseEntity<Temple> create(@RequestBody CreateTempleRequest request,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        Temple created = templeService.createTemple(request, user.getId());
        return ResponseEntity.created(URI.create("/api/temples/" + created.getId())).body(created);
    }

    @PreAuthorize("hasRole('ServiceProvider')")
    @PutMapping("/{id}")
    public ResponseEntity<?> update(@PathVariable Integer id,
            @RequestBody UpdateTempleRequest request,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        Temple updated = templeService.updateTemple(id, request, user.getId());

        if (updated == null) {
            return ResponseEntity.notFound()
                    .build();
        }

        return ResponseEntity.ok(updated);
    }

    @PreAuthorize("hasRole('ServiceProvider')")
    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(@PathVariable Integer id, @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        boolean result = templeService.deleteTemple(id, user.getId());

        if (!result) {
            return ResponseEntity.notFound()
                    .build();
        }

        return ResponseEntity.noContent().build();
    }
}
